src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -450,31 +502,52 @@ protected void insertBufferedRecordsIfFlushTimeThresholdReached() {
    * Invokes insertRows API using the provided offsets which were initially buffered for this
    * partition. This buffer is decided based on the flush time threshold, buffered bytes or number
    * of records
+   *
+   * <p>When the response contains any extra column error, we will attempt to alter the table
+   * accordingly, reopen the channel and re-insert the records
+   *
+   * <p>Rows with error other than that will be added to DLQ if enabled
+   *
+   * @param streamingBufferToInsert the record buffer
    */
-  InsertValidationResponse insertBufferedRecords(StreamingBuffer streamingBufferToInsert) {
-
+  public InsertRowsWithRetryResponse insertBufferedRecordsWithRetry(
+      StreamingBuffer streamingBufferToInsert, int maxInsertAttempt) {
     // intermediate buffer can be empty here if time interval reached but kafka produced no records.
     if (streamingBufferToInsert.isEmpty()) {
       LOGGER.debug("No Rows Buffered for channel:{}, returning", this.getChannelName());
       this.previousFlushTimeStampMs = System.currentTimeMillis();
-      return null;
+      return new InsertRowsWithRetryResponse();
     }
-    InsertValidationResponse response = null;
+    InsertRowsWithRetryResponse response = null;
     try {
-      response = insertRowsWithFallback(streamingBufferToInsert);
+      long lastKafkaOffsetInFailedBuffer = streamingBufferToInsert.getLastOffset();
+      response = insertRowsAndAlterTableWithRetry(streamingBufferToInsert, maxInsertAttempt);
+
       // Updates the flush time (last time we called insertRows API)
       this.previousFlushTimeStampMs = System.currentTimeMillis();
 
       LOGGER.info(
-          "Successfully called insertRows for channel:{}, buffer:{}, insertResponseHasErrors:{}",
+          "Successfully called retry insertRows for channel:{}, buffer:{},"
+              + " insertResponseHasErrors:{}",
           this.getChannelName(),
           streamingBufferToInsert,
           response.hasErrors());
       if (response.hasErrors()) {
-        handleInsertRowsFailures(
-            response.getInsertErrors(), streamingBufferToInsert.getSinkRecords());
+        handleInsertRowsWithRetryFailures(response.insertErrors, response.recordWithError);
       }
-      return response;
+      // the channel was reopened (several times, due to a racing condition in altering table or a
+      // random failure in reinsert the rows), and the channel offset was
+      // reset to be before the current buffer. The offset needs to be reset to be the end of the
+      // buffer to continue ingestion properly
+      // even if the retry does not succeed the reset is necessary to make the behavior consistent
+      // with that without schema evolution
```

Comment:
can you complete the sentence

Reply:
This is actually a complete sentence but I got confused when I look back at it 😆 . Changed